design-proposals/klip-15-new-api-and-client.md

```diff
@@ -0,0 +1,189 @@
+that uses this API.
+
+This KLIP proposes that we:
+
+* Create a ksqlDB client (initially in Java and JavaScript) as the primary interaction point to the event streaming platform
+for an application
+* Create a new server side implementation that supports the client and enables the functionality of the client to be delivered in a straightforward
+and efficient way
+* The creation of a new wire protocol for the client<->server communication.
+* Migrate the functionality of the current HTTP/REST API over to the new server implementation and retire the parts that are no longer needed.
+
+## What is in scope
+
+### The client
+
+* We will create a new client, initially in Java (and potentially in JavaScript). Clients for other languages such as Python and Go will follow later.
+* The client will initially support execution and streaming of queries (both push and pull), inserting of rows into streams, DDL operations and admin operations
+such as list and describe.
+* The client will support a reactive / async interaction style using Java CompletableFuture and Reactive Streams / JDK9 Flow API (or a copy thereof if we cannot upgrade from Java 8)
+* The client will also support a direct / synchronous interaction style.
+
+### The protocol
+
+We will create a new super simple streaming protocol designed to efficiently stream data from server to client or client to server.
+HTTP/REST is a poor fit for streaming data as it is designed for a request/response interaction style. Attempting to implement "streaming"
+over HTTP/REST usually results in a polling approach which is awkward and inefficient compared to a protocol which is designed with streaming in mind from
+the start.
+
+* Binary multiplexed protocol. The protocol will support multiple independent channels over a single connection. We want to discourage clients from opening new
+connections for single operations (e.g. a single query) and closing them again. This leads to poor performance and an inefficient use of resources. Multiplexing allows
+many queries to be running over a single connection. Multiplexing is also important because it is likely each KSQL node will maintain connections to other KSQL nodes
+for clustering purposes (e.g. proxying queries) It is far more efficient for a particular node to hold a single connection (or a small number of connections)
+to each other node rather than creating new connections for each query.
+* Each channel will implement flow control (back pressure) to prevent any one channel overwhelming the connection.
+* The protocol will run over WebSockets - WebSockets clients are available and very easy to use in all languages of interest.
+WebSockets are also usable from browsers - this will be very useful with a JavaScript client.
+* The protocol will also run over TCP. TCP might make sense in simple data centre deployments where there's no particular advantage to using WebSockets.
```

Comment:
typo: data center :)

Reply:
Data centre is how we say it in the UK? I think center is US dialect :)

Reply:
Fair enough